pkg/controller/daemon/util/daemon_util.go

```diff
@@ -0,0 +1,161 @@
+/*
+Copyright 2015 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package daemon
+
+import (
+	"fmt"
+	"strconv"
+
+	"github.com/golang/glog"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/apis/extensions"
+	clientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+	labelsutil "k8s.io/kubernetes/pkg/util/labels"
+	podutil "k8s.io/kubernetes/pkg/util/pod"
+)
+
+const (
+	// RevisionAnnotation is the revision annotation of a daemon set pod template which records its rollout sequence
+	RevisionAnnotation = "daemonset.kubernetes.io/revision"
+	PodTemplateLabel   = "damonset.kubernetes.io/daemon-name"
+)
+
+type podTemplateListFunc func(string, api.ListOptions) (*api.PodTemplateList, error)
+
+func GetOrCreatePodTemplate(ptc PodTemplateControllerInterface, ds *extensions.DaemonSet, c clientset.Interface) (*api.PodTemplate, error) {
+	var podTemplate *api.PodTemplate
+
+	podsTemplates, err := listPodTemplates(ds, c)
+	if err != nil {
+		return nil, err
+	}
+	if len(podsTemplates.Items) == 0 {
+		podTemplate, err = CreatePodTemplateFromDS(ptc, ds, "1")
```

Comment:
Only this comment from @bgrant0607  https://github.com/kubernetes/kubernetes/pull/31693#issuecomment-243675435

I have to say that using it was quite pleasant. Do you see other options to store history?